enforcer/src/main/java/org/wso2/choreo/connect/enforcer/websocket/WebSocketResponseObserver.java

```diff
@@ -0,0 +1,107 @@
+ * the related API, application, subscriber etc
+ */
+public class WebSocketResponseObserver implements StreamObserver<WebSocketFrameRequest> {
+
+    private static final Logger logger = LogManager.getLogger(WebSocketResponseObserver.class);
+    private String apiThrottleKey;
+    private String subscriptionThrottleKey;
+    private String applicationThrottleKey;
+    private final StreamObserver<WebSocketFrameResponse> responseStreamObserver;
+    private final WebSocketHandler webSocketHandler = new WebSocketHandler();
+    private String streamId;
+    private boolean throttleKeysInitiated;
+
+    public WebSocketResponseObserver(StreamObserver<WebSocketFrameResponse> responseStreamObserver) {
+        this.responseStreamObserver = responseStreamObserver;
+    }
+
+    @Override
+    public void onNext(WebSocketFrameRequest webSocketFrameRequest) {
+        logger.debug("Websocket frame received");
+        if (!this.throttleKeysInitiated) {
+            initializeThrottleKeys(webSocketFrameRequest);
+        }
+        WebSocketThrottleResponse webSocketThrottleResponse = webSocketHandler.process(webSocketFrameRequest);
+        if (webSocketThrottleResponse.getWebSocketThrottleState() == WebSocketThrottleState.OK) {
+            WebSocketFrameResponse response = WebSocketFrameResponse.newBuilder().setThrottleState(
+                    WebSocketFrameResponse.Code.OK).build();
+            responseStreamObserver.onNext(response);
+        } else {
+            logger.debug("throttle period" + webSocketThrottleResponse.getThrottlePeriod());
+            WebSocketFrameResponse response = WebSocketFrameResponse.newBuilder().setThrottleState(
+                    WebSocketFrameResponse.Code.OVER_LIMIT).setThrottlePeriod(
+                    webSocketThrottleResponse.getThrottlePeriod()).build();
+            responseStreamObserver.onNext(response);
+        }
+    }
+
+    @Override
+    public void onError(Throwable throwable) {
+        logger.error("websocket metadata service onError: " + throwable.toString());
+        WebSocketFrameService.removeObserver(streamId);
+    }
+
+    @Override
+    public void onCompleted() {
+        WebSocketFrameService.removeObserver(streamId);
+    }
+
+    private void initializeThrottleKeys(WebSocketFrameRequest webSocketFrameRequest) {
+        Map<String, String> extAuthMetadata = webSocketFrameRequest.getMetadata().getExtAuthzMetadataMap();
+        String basePath = extAuthMetadata.get(APIConstants.GW_BASE_PATH_PARAM);
+        String version = extAuthMetadata.get(APIConstants.GW_VERSION_PARAM);
+        String applicationId = extAuthMetadata.get(MetadataConstants.APP_ID);
+        String streamId = extAuthMetadata.get(MetadataConstants.GRPC_STREAM_ID);
+        String apiContext = basePath + ':' + version;
+        this.apiThrottleKey = apiContext;
+        this.subscriptionThrottleKey = applicationId + ":" + apiContext;
+        this.applicationThrottleKey = applicationId;
```

Comment:
application throttle key should have the user name as well

Reply:
Will add the user name. This key is kept as a reference here. The application key used for throttling in defined in the WebSocketThrottleFilter correctly. https://github.com/wso2/product-microgateway/blob/6784b0c44b57aaaca842c1247398f0d2f50e7d87/enforcer/src/main/java/org/wso2/choreo/connect/enforcer/websocket/WebSocketThrottleFilter.java#L130